src/extract_gpuinfo_ascend.c

```diff
@@ -0,0 +1,251 @@
+  sprintf(pdev, "%d-%d", (short)card_id, (short)device_id);
+}
+
+static void _decode_card_device_id_from_pdev(const char *pdev, int *card_id, int *device_id) {
+  sscanf(pdev, "%d-%d", card_id, device_id);
+}
+
+static void gpuinfo_ascend_populate_static_info(struct gpu_info *_gpu_info) {
+  struct gpu_info_ascend *gpu_info = container_of(_gpu_info, struct gpu_info_ascend, base);
+  struct gpuinfo_static_info *static_info = &gpu_info->base.static_info;
+  static_info->integrated_graphics = false;
+  RESET_ALL(static_info->valid);
+
+  int card_id, device_id;
+  _decode_card_device_id_from_pdev(_gpu_info->pdev, &card_id, &device_id);
+
+  struct dcmi_chip_info *chip_info = malloc(sizeof(struct dcmi_chip_info));
+  last_dcmi_return_status = dcmi_get_device_chip_info(card_id, device_id, chip_info);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    // assume Ascend only use ASCII code for chip name
+    strcpy(static_info->device_name, (char *) chip_info->chip_name);
+    SET_VALID(gpuinfo_device_name_valid, static_info->valid);
+  }
+  free(chip_info);
+  // todo: it seems that other static infos are not supported by Ascend DCMI for now, will add if possible in future
+}
+
+static void gpuinfo_ascend_refresh_dynamic_info(struct gpu_info *_gpu_info) {
+  struct gpu_info_ascend *gpu_info = container_of(_gpu_info, struct gpu_info_ascend, base);
+  struct gpuinfo_dynamic_info *dynamic_info = &gpu_info->base.dynamic_info;
+  RESET_ALL(dynamic_info->valid);
+  dynamic_info->encode_decode_shared = false;
+
+  int card_id, device_id;
+  _decode_card_device_id_from_pdev(_gpu_info->pdev, &card_id, &device_id);
+
+  unsigned aicore_freq;
+  last_dcmi_return_status = dcmi_get_device_frequency(card_id, device_id, DCMI_FREQ_AICORE_CURRENT_, &aicore_freq);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    dynamic_info->gpu_clock_speed = aicore_freq;
+    SET_VALID(gpuinfo_gpu_clock_speed_valid, dynamic_info->valid);
+  }
+
+  unsigned aicore_max_freq;
+  last_dcmi_return_status = dcmi_get_device_frequency(card_id, device_id, DCMI_FREQ_AICORE_MAX, &aicore_max_freq);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    dynamic_info->gpu_clock_speed_max = aicore_max_freq;
+    SET_VALID(gpuinfo_gpu_clock_speed_max_valid, dynamic_info->valid);
+  }
+
+  unsigned hbm_freq;
+  last_dcmi_return_status = dcmi_get_device_frequency(card_id, device_id, DCMI_FREQ_HBM, &hbm_freq);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    dynamic_info->mem_clock_speed = hbm_freq;
+    SET_VALID(gpuinfo_mem_clock_speed_valid, dynamic_info->valid);
+  }
+
+  unsigned aicore_util_rate;
+  last_dcmi_return_status = dcmi_get_device_utilization_rate(card_id, device_id, DCMI_UTILIZATION_RATE_AICORE, &aicore_util_rate);
+  if (last_dcmi_return_status == DCMI_SUCCESS) {
+    dynamic_info->gpu_util_rate = aicore_util_rate;
+    SET_VALID(gpuinfo_gpu_util_rate_valid, dynamic_info->valid);
+  }
+
+  struct dsmi_hbm_info_stru *hbm_info = malloc(sizeof(struct dsmi_hbm_info_stru));
```

Comment:
Why malloc when you can have a plain `struct dsmi_hbm_info_stru` instead of a pointer here?

Reply:
Thanks for the suggestion! You're right, use plain struct is simpler and be better.